Implement `is_lazy_import` inspection API

chuanzzx opened the pull request:
# Summary
Add `_importlib.is_lazy_import`. Then, we can import `_importlib` and use its method `is_lazy_import` to check if a key in the specified dict is lazy when using Python.

# Test Plan

We will import `_importlib` and use its function `is_lazy_import` to validate the behavior. The first argument needs to be a dictionary, and the second argument needs to be a string which is the module name you would like to check if it is lazy now.
For example,
```
import _importlib
_importlib.is_lazy_import(globals(), "math")
```

-  It should return `True` when `math` is lazy.
-  It should return `False` when `math` is not lazy.
-  It will produce error message when facing an error.

## Without Lazy Imports
Run 
```
$ ./python.exe
>>> import _importlib
```
### 1. Non-imported module
When we test a non-imported module, it will throw the exception `KeyError`.
This is because the module is not imported.
```
>>> _importlib.is_lazy_import(globals(), "math")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'math'
```

### 2. Imported module
When we test an imported module, it should return `False`.
We are not using Lazy Imports, so the module is eagerly imported.
```
>>> import math
>>> _importlib.is_lazy_import(globals(), "math")
False
```

## With Lazy Imports
Run
```
$ ./python.exe -L
>>> import _importlib
```

### 1. Non-imported module
When we test a non-imported module, it will throw the exception `KeyError`. 
This is because the module is not imported.
```
>>> import _importlib
>>> _importlib.is_lazy_import(globals(), "math")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'math'
```

### 2. Imported module
When we test an imported module, it should return `True`.
We are using Lazy Imports, so the module **should be lazy** until it is used.
```
>>> import math
>>> _importlib.is_lazy_import(globals(), "math")
True
```

### 3. Executed module
After we use `math.pi`, the module is loaded. Thus, it should return `False`.
```
>>> math.pi
3.141592653589793
>>> _importlib.is_lazy_import(globals(), "math")
False
```

### 4. Wrong arguments

#### Wrong numbers of arguments
If we do not give the correct numbers of arguments to `is_lazy_import`, it will return an error message.
```
>>> _importlib.is_lazy_import(globals())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: is_lazy_import expected 2 arguments, got 1
```

#### Wrong dictionary type
If we provide a non-dict type argument in argument 1, it will return an error message.
```
>>> _importlib.is_lazy_import(123, "math")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: is_lazy_import() argument 1 must be dict, not int
```

#### Wrong key type
If we provide a non-string type argument in argument 2, it will return an error message.
```
>>> _importlib.is_lazy_import(globals(), 123)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: is_lazy_import() argument 2 must be str, not int
```

chuanzzx committed:
Add `_lazydebug.is_lazy_key`

Can import `_lazydebug` and use its method `is_lazy_key` to check if a key in the specified dict is lazy.

chuanzzx committed:
Remove debugging logs and add the doc for is_lazy_key_.

itamaro commented on Include/dictobject.h:
```diff
@@ -31,6 +31,7 @@ PyAPI_FUNC(PyObject *) PyDict_Values(PyObject *mp);
 PyAPI_FUNC(PyObject *) PyDict_Items(PyObject *mp);
 PyAPI_FUNC(Py_ssize_t) PyDict_Size(PyObject *mp);
 PyAPI_FUNC(PyObject *) PyDict_Copy(PyObject *mp);
+PyAPI_FUNC(int) PyDict_IsLazyKey(PyObject *mp, PyObject *key);
```
naming bikeshed!

how about `PyDict_IsLazyImport`?

the API we specified in the PEP (PR still pending, https://github.com/python/peps/pull/2613) is `importlib.is_lazy_import`, so I think this name would be consistent with it

I guess another nit is whether this should be a PyDict API or PyImport API

chuanzzx replied:
Changed the naming.

About the position, I think it depends on which we care more in this API, the `dict` or the `key`.

itamaro commented on Modules/Setup:
```diff
@@ -156,6 +156,7 @@ PYTHONPATH=$(COREPYTHONPATH)
 #math mathmodule.c
 #mmap mmapmodule.c
 #select selectmodule.c
+#_lazydebug _imp.c
```
what's this?

chuanzzx replied:
Oops, that is I did to mimic the implementation of `_crypt`, but I found it doesn't work. Let me remove it.

itamaro commented on Modules/_imp.c:
```diff
@@ -0,0 +1,29 @@
+
+static struct PyModuleDef lazydebugmodule = {
+    PyModuleDef_HEAD_INIT,
+    "_lazydebug",
```
we should have this under `_importlib`

related to the other discussion we had - `_imp.c` is not the appropriate name for this module - let's move it to `_importlib.c` and export it as `_importlib`

chuanzzx replied:
Yes, this also would be helpful for the following tasks!

itamaro commented on Objects/dictobject.c:
```diff
@@ -3929,6 +3929,30 @@ static PyMethodDef mapp_methods[] = {
+    PyObject *value = _PyDict_GetItemKeepLazy(mp, key);
+
+    // error
+    if (value == NULL) {
+        return -1;
+    }
```
I think it would be cleaner to throw an exception instead of returning `-1` (e.g. `KeyError`), which would allow the Python wrapper to return `bool` instead if `int`

itamaro reviewed (changes requested):
this looks good!
see inline comments.
we also need to wrap this in `Lib/importlib` as `is_lazy_import()` (that would wrap a call to `_importlib...`

chuanzzx committed:
Update naming, exception, and return value type

1. Change the naming of exported module and internal function.
2. Add exception in `PyDict_IsLazyImport`
3. Change the return value type of `is_lazy_key`

chuanzzx committed:
Retrieve unnecessary doc change

chuanzzx commented:
Updated the code and test plan based on Itamar's feedback. Thanks! :)

itamaro commented on Modules/_importlib.c:
```diff
@@ -0,0 +1,33 @@
+    int res = PyDict_IsLazyImport(dict, key);
+
+    if (res == 1) {
+        return Py_BuildValue("O", Py_True);
```
```suggestion
        Py_RETURN_TRUE;
```

itamaro commented on Modules/_importlib.c:
```diff
@@ -0,0 +1,33 @@
+    if (res == 1) {
+        return Py_BuildValue("O", Py_True);
+    }
+    return Py_BuildValue("O", Py_False);
```
```suggestion
    Py_RETURN_FALSE;
```

itamaro commented on Modules/_importlib.c:
```diff
@@ -0,0 +1,33 @@
+#include "clinic/_importlib.c.h"
+
+static PyObject *
+is_lazy_key_impl(PyObject *module, PyObject *dict, PyObject *key)
```
s/key/import/

itamaro commented on Objects/dictobject.c:
```diff
@@ -3929,6 +3929,33 @@ static PyMethodDef mapp_methods[] = {
+
+    // error
+    if (value == NULL) {
+        PyErr_SetString(PyExc_KeyError,
+                        "cannot find this key "
+                        "in the specified dictionary");
+        return -1;
```
as discussed in chat - either set the exception in the caller instead of here, or change this function to return `PyObject *`

itamaro reviewed (changes requested):
almost there! :)

chuanzzx committed:
Change the KeyError way and function name.

itamaro commented on Objects/dictobject.c:
```diff
@@ -3929,6 +3929,30 @@ static PyMethodDef mapp_methods[] = {
 };
 
+
+// return 1 if the module `key` in `mp` is not loaded (a lazy key)
+// return 0 if the module `key` in `mp` is loaded
+// return -1 if existing an error
```
wording nits
```suggestion
// return 1 if `key` in `mp` contains a lazy import object
// return 0 if `key` in `mp` is not a lazy import object
// return -1 if `key` doesn't exist in `mp`, or an error occurred
```

itamaro commented on Modules/_importlib.c:
```diff
@@ -0,0 +1,39 @@
+        PyErr_SetObject(PyExc_KeyError, key);
+        return NULL;
+    }
+    else if (res == 1) {
+        Py_RETURN_TRUE;
+    }
+    else {
+        Py_RETURN_FALSE;
+    }
```
nit: no need for else's after return
```suggestion
    if (res == 1) {
        Py_RETURN_TRUE;
    }
    Py_RETURN_FALSE;
```

itamaro reviewed (approved):
thanks, this lgtm now!
left a couple of nit comments for your consideration

chuanzzx committed:
Change comments and remove unnecessary else.
